# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [82]:
# import numpy and pandas
import numpy as np
import pandas as pd


# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [83]:
# Run this code:

pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [84]:
# Your code here:

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [85]:
# Your code here:
pokemon.groupby('Legendary').count()['#'].reset_index()

Legendary    #
0      False  735
1       True   65

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [86]:
# Your code here:
pokemon.groupby('Legendary').agg(('mean','std'))['Total']


/var/folders/51/zl81hkbj365fyht29sw2pjsw0000gn/T/ipykernel_41836/2085708270.py:2: FutureWarning: ['Name', 'Type 1', 'Type 2'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  pokemon.groupby('Legendary').agg(('mean','std'))['Total']


mean         std
Legendary                        
False      417.213605  106.760417
True       637.384615   60.937389

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [87]:
legendary=pokemon[(pokemon['Legendary']==True)]

In [88]:
nonlegendary=pokemon[(pokemon['Legendary']==False)]

In [89]:
from statsmodels.stats.weightstats import ttest_ind

In [90]:
print(np.var(legendary['Total']), np.var(nonlegendary['Total']))

3656.236686390535 11382.279542783135


In [91]:
import scipy.stats as stats
stats.ttest_ind(legendary['Total'], nonlegendary['Total'], equal_var=False)

Ttest_indResult(statistic=25.8335743895517, pvalue=9.357954335957446e-47)

In [92]:
 ttest_ind(legendary['Total'], nonlegendary['Total'],usevar='unequal')


(25.8335743895517, 9.357954335957717e-47, 102.79988763435725)

What do you conclude from this test? Write your conclusions below.


**The p-value of the test comes out to be equal to 9.357e-47, which is smaller than the significance level alpha (that is, 0.05). This implies that the average Total of legendary is statistically different from the average Total of non legendary.**

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [93]:
# Your code here:
pokemon.groupby('Type 1').count()['#']


Type 1
Bug          69
Dark         31
Dragon       32
Electric     44
Fairy        17
Fighting     27
Fire         52
Flying        4
Ghost        32
Grass        70
Ground       32
Ice          24
Normal       98
Poison       28
Psychic      57
Rock         44
Steel        27
Water       112
Name: #, dtype: int64

In [94]:
pokemon.head(3)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1  2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2  3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [46]:
# Your code here:
water=pokemon[(pokemon['Type 1']=='Water')][['Name','Total']]
water.describe()


Total
count  112.000000
mean   430.455357
std    113.188266
min    200.000000
25%    328.750000
50%    455.000000
75%    502.250000
max    770.000000

In [50]:
pokemon[['Name','Total']].describe()

Total
count  800.00000
mean   435.10250
std    119.96304
min    180.00000
25%    330.00000
50%    450.00000
75%    515.00000
max    780.00000

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [47]:
nonwater=pokemon[(pokemon['Type 1']!='Water')][['Name','Total']]
nonwater.describe()

Total
count  688.000000
mean   435.859012
std    121.091682
min    180.000000
25%    330.000000
50%    450.000000
75%    515.000000
max    780.000000

In [48]:
# Your code here:
stats.ttest_ind(water['Total'], nonwater['Total'])

Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Write your conclusion below.

**The p-value of the test comes out to be equal to 0.6587, which is greater than the significance level alpha (that is, 0.05). This implies that we can say that the average total of water pokemon is statistically not different from the average total of non water pokemon.** 

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [95]:
# Your code here:
from scipy.stats import ttest_rel
ttest_rel(pokemon['Defense'],pokemon['Attack'],axis=0)

Ttest_relResult(statistic=-4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.


**The P-value is less than 0.05 hence there is a significant difference between them.**


We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [56]:
# Your code here:
stats.ttest_ind(pokemon['Sp. Def'], pokemon['Sp. Atk'])


Ttest_indResult(statistic=-0.6041290031014401, pvalue=0.5458436328840358)

Describe the results of the test in the cell below.


**No. As the P-value being larger than 0.05 illustrates, there is no significant difference between these two.** 


As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [61]:
# Your code here:
from scipy.stats import ttest_1samp
ttest_1samp(pokemon['Defense']-pokemon['Attack'],popmean=0)

Ttest_1sampResult(statistic=-4.325566393330478, pvalue=1.7140303479358558e-05)

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [81]:
tbl=pd.crosstab(index=pokemon['Legendary'], columns=pokemon['Type 1'].eq(('Water')))
tbl

Type 1     False  True
Legendary             
False        627   108
True          61     4

In [72]:
pokemon['water']=np.where(pokemon['Type 1']=='Water',True,False)

In [73]:
# Your code here:
table=pd.crosstab(index=pokemon['Legendary'], columns=pokemon['water'])
table

water      False  True
Legendary             
False        627   108
True          61     4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [80]:
# Your code here:
from scipy.stats import chi2_contingency
chi2_contingency(tbl)


(2.9429200762850503,
 0.08625467249550949,
 1,
 array([[632.1, 102.9],
        [ 55.9,   9.1]]))

Based on a 95% confidence, should we reject the null hypothesis?


**Since the p value (0.08625) is greater than alpha (0.05), we accept the null hypothesis (H0) which states that there is no significant relationship between the variables Legendary and Water Type.**
